norns/engine-study-3.md

```diff
@@ -0,0 +1,1489 @@
+		Out.ar(\outSend2.kr, (snd * \levelSend2.kr(0)).dup);
+	}).add;
+
+	// define our delay:
+	SynthDef("delay", {
+		Out.ar(\out.kr, CombC.ar(In.ar(\in.kr, 2),1.0,0.2,3.2));
+	}).add;
+
+	// define our reverb:
+	SynthDef("reverb", {
+		var sig = In.ar(\in.kr, 2);
+		Out.ar(\out.kr, FreeVerb2.ar(sig[0], sig[1], 1.0, 0.7, 0.2, 1.5));
+	}).add;
+
+	// define our main output:
+	SynthDef("main", {
+		Out.ar(\out.kr, In.ar(\in.kr, 2));
+	}).add;
+
+	// we sync the Server here so that the common SynthDefs above
+	//   are present on the Server when requested below
+	s.sync;
+
+	// build our source and pass it arguments:
+	~synths[\source] = Synth.new("source", [
+		\outMain, ~busses[\mainOut], // connecting to the mainOut bus
+		\outSend1, ~busses[\delaySend], // connecting to the delaySend bus
+		\outSend2, ~busses[\reverbSend] // connecting to the reverbSend bus
+	]);
+
+	// build our delay AFTER our source
+	//   and pass it arguments:
+	~synths[\delay] = Synth.after(~synths[\source], "delay", [
+		\in, ~busses[\delaySend], // input = the delaySend bus
+		\out, ~busses[\mainOut] // output = the mainOut bus
+	]);
+
+	// build our reverb AFTER our delay
+	//   and pass it arguments:
+	~synths[\reverb] = Synth.after(~synths[\delay], "reverb", [
+		\in, ~busses[\reverbSend], // input = the reverbSend bus
+		\out, ~busses[\mainOut] // output = the mainOut bus
+	]);
+
+	// build our main output AFTER our reverb
+	//   and pass it arguments:
+	~synths[\main] = Synth.after(~synths[\reverb], "main", [
+		\in, ~busses[\mainOut], // input = the mainOut bus
+		\out, 0 // output = the default output device
+	]);
+}.play;
+)
+```
+</details>
+
+After executing the code, you should hear a regular plucky note at 330 Hz. We can control its send levels via these commands:
+
+```js
+// send to main:
+~synths[\source].set(\levelMain,1);
+// send to delay:
+~synths[\source].set(\levelSend1,1);
+// send to reverb:
+~synths[\source].set(\levelSend2,1);
+// don't send to main:
+~synths[\source].set(\levelMain,0);
+// don't send to delay:
+~synths[\source].set(\levelSend1,0);
+// don't send to reverb:
+~synths[\source].set(\levelSend2,0);
+```
+
+### Class file
+
+It'd be nice to extend our sketch by:
+
+- building it into a full Class file
+- using [Groups](https://doc.sccode.org/Classes/Group.html) to organize our synths / nodes
+- adding panning to each send (dry source, delay, reverb)
+- establishing a few commands for control over level, panning, and the Hertz of the played note
+
+Of note in this example:
+
+- rather than force our source sound into stereo with `.dup` (as in our previous example), we'll simply feed it into a two channel panning UGen
+- we'll use a Group (assigned var `g`) to control the order of execution, which will include specifying `target:g` and `addAction:\addToTail` for each of our synths
+- we'll use `s.sync` every time we want to execute what is queued for the Server, before adding more synths / nodes
```

Comment:
not quite accurate: `sync` pauses the enclosing thread until all pending events are done; it doesn't synchronize the excecution of events on the server (there is no "queue" as described)